File: algorithms/[1006]笨阶乘/solution.ts
// topics = ["数学"]

function clumsy(N: number): number {
  if (N == 1) {
    return 1;
  }
  if (N == 2) {
    return 2;
  }
  if (N == 3) {
    return 6;
  }
  if (N == 4) {
    return 7;
  }

  // 当 N > 4 时，可以根据 N 对 4 取模的余数进行分类讨论
  if (N % 4 == 0) {
    return N + 1;
  }
  if (N % 4 == 3) {
    return N - 1;
  }
  return N + 2;
}

File: algorithms/[1006]笨阶乘/solution.py
# topics = ["栈"]


class Solution:
    def clumsy(self, N: int) -> int:
        """
        栈模拟
        """
        # 当前计算结果
        curr: int = N
        # 保留之前计算结果，只需要栈顶元素
        last: int = 0
        # 0 - 乘法；1 - 除法；2 - 加法；3 - 减法
        op = 0

        for i in range(1, N)[::-1]:
            if op == 0:
                curr *= i
            elif op == 1:
                curr = int(curr / i)
            elif op == 2:
                curr += i
            else:
                last += curr
                curr = -i

            op = 0 if op == 3 else op + 1

        return last + curr
